### PR TITLE
Spec optional conversion to RGB in VideoFrame.copyTo()

### PR DESCRIPTION
**Pull request by Djuffin**
This change should address the [issue](https://github.com/w3c/webcodecs/issues/92) of pixel format conversion in `VideoFrame.copyTo()`.

***
<a href="https://pr-preview.s3.amazonaws.com/Djuffin/webcodecs/pull/754.html" title="Last updated on Apr 16, 2024, 8:43 AM UTC (0df4aee)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/webcodecs/754/871ccb3...Djuffin:0df4aee.html" title="Last updated on Apr 16, 2024, 8:43 AM UTC (0df4aee)">Diff</a>

### CODE REVIEW

**Comment by Djuffin** [comment]
There are probably some missing details, typos and omissions.
But I'd love to get some feedback on the general approach in this change.

File: index.src.html [hunk 1]
```diff
@@ -3728,8 +3730,8 @@
 
 : <dfn method for=VideoFrame>copyTo(|destination|, |options|)</dfn>
 :: Asynchronously copies the planes of this frame into |destination| according
-    to |options|. The format of the data is the same as this {{VideoFrame}}'s
-    {{VideoFrame/format}}.
+    to |options|. The format of the data is specified by {{VideoFrameCopyToOptions/format}},
+    if present, or the same as {{VideoFrame/format}}.
```
Comment (sandersdan) [inline index.src.html:3734]: Recommend replacing "or" with "otherwise".

File: index.src.html [hunk 2]
```diff
@@ -4166,10 +4185,36 @@
...
+        on |canvas| with `"2d"` and |canvasSettings|
+    3. Call {{CanvasDrawImage/drawImage()}} on |context| with
+        `rect.x`, `rect.y`, `rect.width`, `rect.height`, `0`, `0`, `rect.width`,
+        `rect.height`.
+    4. Let |imageData| be the result of calling {{CanvasImageData/getImageData()}} with
+        `0`, `0`, |rect.width|, |rect.height|.
+    5. If |format| is equal to {{BGRA}} or {{BGRX}}:
+        1. Let |pixelIndex| be `0`
+        2. While |pixelIndex| is less than `rect.width` multiplied by `rect.height`:
+            1. Swap items in |imageData|.{{ImageData/data}} at indexes `4 * pixelIndex`
```
Comment (sandersdan) [inline index.src.html:4203]: Nit: `items` is perhaps unclear, recommend "elements".

File: index.src.html [hunk 3]
```diff
@@ -4256,6 +4303,16 @@
     to specify an offset and stride for each plane in the destination
     {{BufferSource}}. If unspecified, the planes will be tightly packed. It is
     invalid to specify planes that overlap.
+: <dfn dict-member for=VideoFrameCopyToOptions>format</dfn>
+:: A {{VideoPixelFormat}} for the pixel data in the destination
+    {{BufferSource}}. Explicity values cab be:
+    {{RGBA}}, {{RGBX}}, {{BGRA}}, {{BGRX}} and the same as {{VideoFrame/format}}.
```
Comment (sandersdan) [inline index.src.html:4309]: What happens if `frame.format` is `null`?

File: index.src.html [hunk 4]
```diff
@@ -4256,6 +4303,16 @@
...
+    If unspecified, the {{VideoFrame/format}} is used.
+: <dfn dict-member for=VideoFrameCopyToOptions>colorSpace</dfn>
+:: A {{PredefinedColorSpace}} that <em class="rfc2119">SHALL</em> be used to as
+    a target color space for for the pixel data in the destination
+    {{BufferSource}}.
```
Comment (sandersdan) [inline index.src.html:4314]: What happens if this member is not present when doing a conversion?

What happens if it is present when not doing a conversion?

File: index.src.html (hunk 1 above)
Comment (aboba) [reply index.src.html:3734]: "or the same as" -> "otherwise"

File: index.src.html (hunk 3 above)
Comment (aboba) [reply index.src.html:4309]: "Explicity values cab" -> "Potential values are"
"and the same as" -> "and other values of"

File: index.src.html (hunk 4 above)
Comment (Djuffin) [reply index.src.html]: I added explanation around it.

File: index.src.html [hunk 5]
```diff
@@ -4166,10 +4191,36 @@
...
+        {{CanvasRenderingContext2DSettings/colorSpace}} set to |colorSpace|.
+    3. Let |context| be the result of calling {{OffscreenCanvas/getContext()}}
+        on |canvas| with `"2d"` and |canvasSettings|
+    3. Call {{CanvasDrawImage/drawImage()}} on |context| with
+        `rect.x`, `rect.y`, `rect.width`, `rect.height`, `0`, `0`, `rect.width`,
+        `rect.height`.
+    4. Let |imageData| be the result of calling {{CanvasImageData/getImageData()}} with
+        `0`, `0`, |rect.width|, |rect.height|.
+    5. If |format| is equal to {{BGRA}} or {{BGRX}}:
+        1. Let |pixelIndex| be `0`
```
Comment (chrisn) [inline index.src.html:4207]:
```suggestion
        1. Let |pixelIndex| be `0`.
```

File: index.src.html [hunk 6]
```diff
@@ -4166,10 +4191,36 @@
...
+        2. While |pixelIndex| is less than `rect.width` multiplied by `rect.height`:
+            1. Swap elements in |imageData|.{{ImageData/data}} at indexes `4 * pixelIndex`
+                 and `4 * pixelIndex + 2`.
+            2. Increment |pixelIndex| by `1`
```
Comment (chrisn) [inline index.src.html:4211]:
```suggestion
            2. Increment |pixelIndex| by `1`.
```

File: index.src.html [hunk 7]
```diff
@@ -4166,10 +4191,36 @@
...
+    6. Let |init| be a new {{VideoFrameBufferInit}} constructed with
+        {{VideoFrameBufferInit/format}} = |format|,
+        {{VideoFrameBufferInit/codedWidth}} = `rect.width`,
+        {{VideoFrameBufferInit/codedHeight}} = `rect.height`
```
Comment (chrisn) [inline index.src.html:4215]:
```suggestion
        {{VideoFrameBufferInit/codedHeight}} = `rect.height`.
```

File: index.src.html [hunk 8]
```diff
@@ -4166,10 +4191,36 @@
...
+    7. Let |convertedFrame| be a new {{VideoFrame}} constructed with
+        |imageData|.{{ImageData/data}} and |init|
```
Comment (chrisn) [inline index.src.html:4217]:
```suggestion
        |imageData|.{{ImageData/data}} and |init|.
```

File: index.src.html [hunk 9]
```diff
@@ -4166,10 +4191,36 @@
...
+    8. Return |convertedFrame|
```
Comment (chrisn) [inline index.src.html:4218]:
```suggestion
    8. Return |convertedFrame|.
```

File: index.src.html [hunk 10]
```diff
@@ -4222,13 +4273,15 @@
 
 VideoFrame CopyTo() Options {#videoframe-copyto-options}
 ------------------------------------------------------------
-Options to specify a rectangle of pixels to copy and the offset and stride of
-planes in the destination buffer.
+Options to specify a rectangle of pixels to copy, their format and the offset
```
Comment (chrisn) [inline index.src.html:4276]:
```suggestion
Options to specify a rectangle of pixels to copy, their format, and the offset
```

File: index.src.html [hunk 11]
```diff
@@ -4256,6 +4309,18 @@
     to specify an offset and stride for each plane in the destination
     {{BufferSource}}. If unspecified, the planes will be tightly packed. It is
     invalid to specify planes that overlap.
+: <dfn dict-member for=VideoFrameCopyToOptions>format</dfn>
+:: A {{VideoPixelFormat}} for the pixel data in the destination
+    {{BufferSource}}. Potential values are:
+    {{RGBA}}, {{RGBX}}, {{BGRA}}, {{BGRX}} and values of {{VideoFrame/format}}.
+    If unspecified or `null`, the {{VideoFrame/format}} is used.
+: <dfn dict-member for=VideoFrameCopyToOptions>colorSpace</dfn>
+:: A {{PredefinedColorSpace}} that <em class="rfc2119">SHALL</em> be used to as
```
Comment (chrisn) [inline index.src.html:4318]:
```suggestion
:: A {{PredefinedColorSpace}} that <em class="rfc2119">SHALL</em> be used as
```

File: index.src.html [hunk 12]
```diff
@@ -4256,6 +4309,18 @@
...
+    a target color space for for the pixel data in the destination
```
Comment (chrisn) [inline index.src.html:4319]:
```suggestion
    a target color space for the pixel data in the destination
```

**Review by chrisn — CHANGES_REQUESTED** [review]

File: index.src.html [hunk 13]
```diff
@@ -4256,6 +4309,18 @@
...
+: <dfn dict-member for=VideoFrameCopyToOptions>format</dfn>
+:: A {{VideoPixelFormat}} for the pixel data in the destination
+    {{BufferSource}}. Potential values are:
+    {{RGBA}}, {{RGBX}}, {{BGRA}}, {{BGRX}} and values of {{VideoFrame/format}}.
+    If unspecified or `null`, the {{VideoFrame/format}} is used.
+: <dfn dict-member for=VideoFrameCopyToOptions>colorSpace</dfn>
+:: A {{PredefinedColorSpace}} that <em class="rfc2119">SHALL</em> be used as
+    a target color space for the pixel data in the destination
+    {{BufferSource}}, but only if {{VideoFrameCopyToOptions/format}} is one of
+    {{RGBA}}, {{RGBX}}, {{BGRA}}, {{BGRX}}, otherwise it is ignored.
```
Comment (sandersdan) [inline index.src.html:4321]: It's also necessary that format differs from the frame format, otherwise the result will be in the frame's color space.

We should probably enforce that the color space is only set in the conversion case, because some day we might want to add color conversion to the other paths as well.

Comment (Djuffin) [reply index.src.html:4335]: Good point. I rewrote it so if `format` is explicitly set it's always a conversion case.

**Comment by Djuffin** [comment]
@padenot Could you please take a look? I'd like to merge this PR some time soon.

File: index.src.html [hunk 14]
```diff
@@ -4179,10 +4205,36 @@
             [=combined buffer layout/allocationSize=].
     9. Return |combinedLayout|.
 
+: <dfn for=VideoFrame>Convert to RGB frame</dfn> (with |format|, |rect| and |colorSpace|)
+ :: 1. Let |canvas| be a new {{OffscreenCanvas}} constructed with
+        |rect|.|width| and |rect|.|height|.
```
Comment (padenot) [inline index.src.html:4210]: This isn't something that we can do in spec, using other specs directly like that. All the objects you use can have been overridden in the global this algorithm is running in.

We should use internal stops for canvas instead.

File: index.src.html [hunk 15]
```diff
@@ -4269,6 +4323,18 @@
     to specify an offset and stride for each plane in the destination
     {{BufferSource}}. If unspecified, the planes will be tightly packed. It is
     invalid to specify planes that overlap.
+: <dfn dict-member for=VideoFrameCopyToOptions>format</dfn>
+:: A {{VideoPixelFormat}} for the pixel data in the destination
+    {{BufferSource}}. Potential values are: {{RGBA}}, {{RGBX}}, {{BGRA}},
+    {{BGRX}}. If unspecified or `null`, the the destination
+    {{BufferSource}} will match {{VideoFrame/format}} .
+: <dfn dict-member for=VideoFrameCopyToOptions>colorSpace</dfn>
+:: A {{PredefinedColorSpace}} that <em class="rfc2119">SHALL</em> be used as
```
Comment (padenot) [inline index.src.html:4332]: Why not must?

**Review by padenot — CHANGES_REQUESTED** [review]

File: index.src.html (hunk 14 above)
Comment (Djuffin) [reply index.src.html:4210]: I'm probably missing something here. Could you elaborate a bit?

Isn't it what we do in the Rendering section?

referring to [CanvasDrawImage](https://html.spec.whatwg.org/multipage/canvas.html#canvasdrawimage) [drawImage()](https://html.spec.whatwg.org/multipage/canvas.html#dom-context-2d-drawimage)

Comment (padenot) [reply index.src.html:4210]: No, that's unrelated, we reference it in your link. What happens when `drawImage` is called with a `VideoFrame` is well-defined there. Here you're _using_ another Web Platform object from the innards of our spec.

e.g. what happens if in the document we're doing this conversion, I've maliciously done:

```js
window.OffscreenCanvas.prototype = function() { alert("what now"); };
```

before anything from our spec run?

Comment (Djuffin) [reply index.src.html:4210]: This algorithm does not refer to any particular js code. It explicitly says what object needs to be created by the user agent. That's why I don't think that global overrides should concern us here. With a similar approach half of this spec can be put into question with code like 

```
window.VideoFrame.prototype = function() { alert("what now"); };
```

File: index.src.html (hunk 15 above)
Comment (Djuffin) [reply index.src.html]: done

File: index.src.html (hunk 14 above)
Comment (Djuffin) [reply index.src.html:4210]: I added a note clarifying that UAs don't have to do it via literally creating an OffscreenCanvas as long as the result is the same. I hope this eliminates possible confusion. 

If you still think that the wording leaves room for misinterpretation, I'm very much open to suggestions how to rewrite it.

Comment (padenot) [reply index.src.html:4210]: This is not about "interpretation", it's simply not something we do in specs, and this needs to be rewritten with an abstract operation exported from the canvas code (if that's what you want to use), and calling into this abstract operation.

See https://github.com/whatwg/html/pull/5510#pullrequestreview-411744788, for a reference of this happening, and https://github.com/w3c/IntersectionObserver/issues/427 for the fix.

Comment (Djuffin) [reply index.src.html:4210]: Thanks for the links!

I've noticed that [drawImage()](https://html.spec.whatwg.org/multipage/canvas.html#drawing-images) (step 6) does not specify any details about the rendering (sampling, color conversion etc). It's only saying that "The user agent may use any filtering algorithm". 
So there isn't really an algorithm to extract and call. 
I will have to rewrite it in a more abstract manner, similar to the text for `drawImage()`

Comment (padenot) [reply index.src.html:4210]: We can add an export to something hand-wavy from canvas now, and merge what we want to do here, and then fix the canvas spec as well, your call.

**Review by chrisn — APPROVED** [review]
Looks good, but let's resolve the discussion on overridden globals (I don't mind if that's here or in a follow up PR).

File: index.src.html (hunk 14 above)
Comment (Djuffin) [reply index.src.html:4210]: I [asked](https://github.com/whatwg/html/issues/10210) HTML spec owners about the best way to reuse canvas pixel format conversion algorithm.

File: index.src.html [hunk 16]
```diff
@@ -3764,10 +3767,22 @@
...
+          1. Let |newOptions| be the result of running the <a>Clone Configuration</a>
+               algorithm with |options|.
+          2. Assign `undefined` to |newOptions|.{{VideoFrameCopyToOptions/format}}.
+          3. Let |rgbFrame| be the result of running the [=VideoFrame/Convert to RGB frame=]
+               algorithm with [=this=], |options|.{{VideoFrameCopyToOptions/format}},
+               and |options|.{{VideoFrameCopyToOptions/colorSpace}}.
+          4. Return the result of calling {{VideoFrame/copyTo()}} on |rgbFrame| with
+               |destination| and |newOptions|.
+    7. Otherwise, if |options|.{{VideoFrameCopyToOptions/format}} is not `undefined` or `null`
+        throw a {{NotSupportedError}} {{DOMException}}.
```
Comment (Kaiido) [inline index.src.html:3781]: This also affects the prior steps 1 & 2, but I doubt you can throw directly here. A method that returns a Promise is supposed to return a rejected Promise instead of throwing (like is done in steps 4 and 5).

Comment (Djuffin) [reply index.src.html]: Good point. Thanks

File: index.src.html (hunk 14 above)
Comment (Djuffin) [reply index.src.html]: I've looked at the rest of the WebCodecs spec. It's pretty light on the details when it comes to format and color spaces conversions. (for example [Rendering](https://www.w3.org/TR/webcodecs/#videoframe-rendering) and [AudioData.copyTo](https://www.w3.org/TR/webcodecs/#dom-audiodata-copyto) )

So I rewrote "Convert to RGB frame" in the similar manner. 
Please take another look.

Comment (Djuffin) [reply index.src.html]: @padenot gentle ping

File: index.src.html [hunk 17]
```diff
@@ -3745,29 +3748,41 @@
 
 : <dfn method for=VideoFrame>copyTo(|destination|, |options|)</dfn>
 :: Asynchronously copies the planes of this frame into |destination| according
-    to |options|. The format of the data is the same as this {{VideoFrame}}'s
-    {{VideoFrame/format}}.
+    to |options|. The format of the data is specified by {{VideoFrameCopyToOptions/format}},
+    if present, otherwise {{VideoFrame/format}}.
```
Comment (padenot) [inline index.src.html:3752]: When looking at how this is rendered in HTML, I find it very confusing, can you rephrase it or make it more explicit or something?

File: index.src.html [hunk 18]
```diff
@@ -3745,29 +3748,41 @@
...
+          1. Let |newOptions| be the result of running the <a>Clone Configuration</a>
+               algorithm with |options|.
+          2. Assign `undefined` to |newOptions|.{{VideoFrameCopyToOptions/format}}.
+          3. Let |rgbFrame| be the result of running the [=VideoFrame/Convert to RGB frame=]
+               algorithm with [=this=], |options|.{{VideoFrameCopyToOptions/format}},
+               and |options|.{{VideoFrameCopyToOptions/colorSpace}}.
+          4. Return the result of calling {{VideoFrame/copyTo()}} on |rgbFrame| with
+               |destination| and |newOptions|.
+    7. Otherwise, if |options|.{{VideoFrameCopyToOptions/format}} is not `undefined` or `null`
+        return a promise rejected with a {{NotSupportedError}} {{DOMException}}.
```
Comment (padenot) [inline index.src.html:3781]: We can use ["exists"](https://infra.spec.whatwg.org/#map-exists) (often used for `dictionaries`) here but we should also check for other invalid values. But isn't that duplicate work with "Parse VideoFrameCopyToOptions" invoked above? You've modified it appropriately.

File: index.src.html [hunk 19]
```diff
@@ -4048,10 +4063,17 @@
     6. Let |optLayout| be `undefined`.
     7. If |options|.{{VideoFrameCopyToOptions/layout}} [=map/exists=], assign
         its value to |optLayout|.
-    8. Let |combinedLayout| be the result of running the [=VideoFrame/Compute
-        Layout and Allocation Size=] algorithm with |parsedRect|,
-        {{VideoFrame/[[format]]}}, and |optLayout|.
-    9. Return |combinedLayout|.
+    8. Let |format| be `undefined`.
+    9. If |options|.{{VideoFrameCopyToOptions/format}} is `undefined`,
+        assign {{VideoFrame/[[format]]}} to |format|.
```
Comment (padenot) [inline index.src.html:4068]: Exists again here.

File: index.src.html [hunk 20]
```diff
@@ -4183,10 +4205,34 @@
...
+            {{VideoFrame/[[visible top]]}}, {{VideoFrame/[[visible width]]}},
+            and {{VideoFrame/[[visible height]]}} respectively.
+	      6. Assign |frame|'s {{VideoFrame/[[duration]]}} and |frame|'s
+            {{VideoFrame/[[timestamp]]}} to {{VideoFrame/[[duration]]}} and
+            {{VideoFrame/[[timestamp]]}} respectively.
+        7. Assign |colorSpace| to {{VideoFrame/[[color space]]}}.
+        8. Let |resource| be a new [=media resource=] containing the result of
+            conversion of [=media resource=] referenced by |frame|'s
+            {{VideoFrame/[[resource reference]]}} into a color space and pixel
+            format specified by |colorSpace| and |format| respectively.
```
Comment (padenot) [inline index.src.html:4228]: As good as it can be for now I guess, but thanks for checking everything on HTML and all.

File: index.src.html [hunk 21]
```diff
@@ -4273,6 +4321,18 @@
...
+:: A {{VideoPixelFormat}} for the pixel data in the destination
+    {{BufferSource}}. Potential values are: {{RGBA}}, {{RGBX}}, {{BGRA}},
+    {{BGRX}}. If unspecified or `null`, the the destination
+    {{BufferSource}} will match {{VideoFrame/format}} .
+: <dfn dict-member for=VideoFrameCopyToOptions>colorSpace</dfn>
+:: A {{PredefinedColorSpace}} that <em class="rfc2119">MUST</em> be used as
+    a target color space for the pixel data in the destination
+    {{BufferSource}}, but only if {{VideoFrameCopyToOptions/format}} is one of
+    {{RGBA}}, {{RGBX}}, {{BGRA}}, {{BGRX}}, otherwise it is ignored.
+    If unspecified or `null`, {{srgb}} is used.
```
Comment (padenot) [inline index.src.html:4334]: Exists

File: index.src.html [hunk 22]
```diff
@@ -4273,6 +4321,18 @@
     to specify an offset and stride for each plane in the destination
     {{BufferSource}}. If unspecified, the planes will be tightly packed. It is
     invalid to specify planes that overlap.
+: <dfn dict-member for=VideoFrameCopyToOptions>format</dfn>
+:: A {{VideoPixelFormat}} for the pixel data in the destination
+    {{BufferSource}}. Potential values are: {{RGBA}}, {{RGBX}}, {{BGRA}},
+    {{BGRX}}. If unspecified or `null`, the the destination
```
Comment (padenot) [inline index.src.html:4327]: Exists

File: index.src.html [hunk 23]
```diff
@@ -4273,6 +4321,18 @@
...
+    {{BufferSource}} will match {{VideoFrame/format}} .
```
Comment (padenot) [inline index.src.html:4328]: "will be in the same format as the source", I think matching is fairly vague.

File: index.src.html (hunk 18 above)
Comment (Djuffin) [reply index.src.html:3781]: yeah, you're right we don't need this step

File: index.src.html [hunk 24]
```diff
@@ -4324,14 +4322,14 @@
 : <dfn dict-member for=VideoFrameCopyToOptions>format</dfn>
 :: A {{VideoPixelFormat}} for the pixel data in the destination
     {{BufferSource}}. Potential values are: {{RGBA}}, {{RGBX}}, {{BGRA}},
-    {{BGRX}}. If unspecified or `null`, the the destination
-    {{BufferSource}} will match {{VideoFrame/format}} .
+    {{BGRX}}. If it does not [=map/exist=], the the destination
```
Comment (Kaiido) [inline index.src.html:4325]: nit: There is a double "the" here that you may want to remove?

File: index.src.html (hunk 20 above)
Comment (Djuffin) [reply index.src.html:4226]: Sure, thanks for pointing out that my approach with getContext/drawImage unconventional

**Comment by Djuffin** [comment]
@padenot PTAL

**Review by padenot — APPROVED** [review]
All good, thanks!

**Comment by Djuffin** [comment]
@aboba PTAL and merge if you don't have new comments.
